zngur-def/src/merge.rs

```diff
@@ -0,0 +1,242 @@
+    }
+}
+
+impl<T: Merge> Merge for Option<T> {
+    /// Writes the partial union of `self` and `into` to the latter.
+    ///
+    /// If both `self` and `into` are Some, the underlying values are merged.
+    /// Otherwise, the result is whichever value is Some, or None if neither is.
+    fn merge(self, into: &mut Self) -> MergeResult {
+        match self {
+            Some(src) => match into.as_mut() {
+                Some(dst) => src.merge(dst),
+                None => {
+                    *into = Some(src);
+                    Ok(())
+                }
+            },
+            None => Ok(()),
+        }
+    }
+}
+
+impl<K, V, I: IntoIterator<Item = (K, V)>> Merge<std::collections::HashMap<K, V>> for I
+where
+    K: Eq + std::hash::Hash,
+    V: Merge,
+{
+    /// Merges a sequence of key-value pairs into a hash map.
+    ///
+    /// If a key is present in both `self` and `into`, the corresponding values are merged.
+    /// Otherwise, the entry from `self` is inserted into `into`.
+    ///
+    /// This implementation implies `std::collections::HashMap<K,V>` is `Merge` for all `V: Merge`,
+    /// because HashMap is `IntoIterator`. We use `IntoIterator` to allow literal sequences of
+    /// key-value pairs to be merged into a map.
+    fn merge(self, into: &mut std::collections::HashMap<K, V>) -> MergeResult {
+        for (key, value) in self {
+            match into.entry(key) {
+                std::collections::hash_map::Entry::Vacant(e) => {
+                    e.insert(value);
+                }
+                std::collections::hash_map::Entry::Occupied(mut e) => {
+                    match value.merge(e.get_mut()) {
+                        Ok(()) => {}
+                        Err(message) => {
+                            return Err(message);
+                        }
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Merge for ZngurType {
+    /// Writes the partial union of `self` and `into` to the latter.
+    ///
+    /// PRECONDITION: `self.ty == into.ty`.
+    fn merge(self, into: &mut Self) -> MergeResult {
+        if self.ty != into.ty {
+            panic!(
+                "Attempt to merge different types: {} and {}",
+                self.ty, into.ty
+            );
+        }
+
+        if self.layout != into.layout {
+            return Err(MergeFailure::Conflict(
+                "Duplicate layout policy found".to_string(),
+            ));
+        }
+
+        if self.cpp_ref.is_some() && into.layout != LayoutPolicy::ZERO_SIZED_TYPE {
+            return Err(MergeFailure::Conflict(
+                "cpp_ref implies a zero sized stack allocated type".to_string(),
+            ));
+        }
+
+        self.cpp_value.merge(&mut into.cpp_value)?;
+        self.cpp_ref.merge(&mut into.cpp_ref)?;
+
+        inplace_union(self.wellknown_traits, &mut into.wellknown_traits);
+        inplace_union(self.methods, &mut into.methods);
```

Comment:
I think we should raise error if there is a method with different signatures, though one of the `.zng` files is wrong on its own, so it is not much important.